### PR TITLE
fix: avoid leaking wrapper space from doc comment code block formatting

### LINKED ISSUE #6832: Trailing space on doc comment
## Summary

I tried to format this code:

```rust
use std::{borrow::Cow, hash::Hash};

/// A CSS-like string value.
///
/// Used for `font-family`, `content` (pseudo-elements), custom identifiers,
/// and any other property that takes a textual value.
///
/// # Zero-copy for static strings
///
/// The inner `Cow<'static, str>` means proc_macro-generated code like:
/// ```rust
/// let font = Str::from_static("JetBrains Mono");
/// ```
/// involves **zero heap allocation** — the slice lives in the binary's
/// read-only data segment. Runtime-computed strings fall back to
/// [`Str::from_string`] which heap-allocates via `Cow::Owned`.
///
/// Two `Str` values are equal if their **contents** are equal, regardless
/// of whether one is borrowed and the other owned.
///
/// # Examples
///
/// ```rust
/// use oxidui_style::str::Str;
/// let a: Str = "monospace".into(); // static borrow, no alloc
/// let b = Str::from_string(format!("Font-{}", 42)); // heap-allocated
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Hash, Default)]
pub struct Str(pub Cow<'static, str>);
```

### Expected behavior
I expect no change

### Actual behavior
Add a trailing whitespace on the `/// let font = Str::from_static("JetBrains Mono");` line 
And then fail due to `error_on_unformated=true`
```rust
/// ```rust
/// let font = Str::from_static("JetBrains Mono"); 
/// ```
```

## Configuration
rustfmt.toml
```toml
# Global config
edition = "2024"
style_edition = "2024"
unstable_features = true
newline_style = "Unix"

# Comment / Raw string formatting
error_on_unformatted = true
wrap_comments = true
format_strings = true
comment_width = 80
normalize_comments = true
format_code_in_doc_comments = true
doc_comment_code_block_width = 80
normalize_doc_attributes = true

# code formatting config
error_on_line_overflow = true
max_width = 80

# Function / Struct / Enum / Trait formatting
brace_style = "SameLineWhere"
empty_item_single_line = true
fn_single_line = true
fn_params_layout = "Tall"
reorder_impl_items = true
type_punctuation_density = "Wide"
use_field_init_shorthand = true
overflow_delimited_expr = true

# Macro formatting
merge_derives = false

# Imports / Modules formatting
imports_layout = "HorizontalVertical"
imports_granularity = "Crate"
group_imports = "StdExternalCrate"
reorder_imports = true
reorder_modules = true

# Control flow
control_brace_style = "AlwaysSameLine"
match_arm_blocks = true
match_arm_indent = true
match_block_trailing_comma = true
force_multiline_blocks = true

# Other
condense_wildcard_suffixes = true
hex_literal_case = "Upper"
tab_spaces = 4
trailing_comma = "Vertical"
trailing_semicolon = true
use_try_shorthand = true
float_literal_trailing_zero = "IfNoPostfix"
```

`rustfmt` cli options used (if applicable):

```bash
cargo +nightly fmt
```

## Reproduction Steps

1. Copy this code, and that config, and try to format it

## Meta

`cargo +nighlty fmt -- --version`:
```
rustfmt 1.9.0-nightly (2d76d9bc76 2026-03-09)
```

### PR DESCRIPTION
**Pull request by caelyreth**
fixes #6832

### Summary

Fixes a bug in doc comment code block formatting where `rustfmt` can introduce a trailing space and then report left behind trailing whitespace (if `error_on_unformatted = true`).

This happens with the following configuration:

```toml
format_code_in_doc_comments = true
fn_single_line = true
```

### Fix

Handle the single-line wrapped case when extracting the formatted code block body so wrapper-added whitespace is not preserved in the returned snippet.

The current fix is narrowly scoped to catch only single-line functions in doc comments. A simpler solution would be to wrap the final result with `remove_trailing_white_spaces()`, but I don’t think it’s ideal since it would somehow impact more cases.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -382,6 +382,22 @@ fn format_code_block(
...
+        .trim_end_matches('\n')
+        .strip_prefix(FN_MAIN_PREFIX.trim_end())
+        .and_then(|s| s.strip_prefix(' '))
+        .and_then(|s| s.strip_suffix(" }"))
+    {
+        return Some(FormattedSnippet {
+            snippet: body.to_owned(),
+            non_formatted_ranges: formatted.non_formatted_ranges,
+        });
+    }
```
Comment (matthewhughes934) [inline src/lib.rs]: This means we have two sets of logic for unindenting the block: the added one here, and the original logic below. I think it's best to keep that as a single piece of logic.

I also don't think we need to special case formatting as a single line, and instead just always trim space before the added `}`. Both of these could be done by updating the end of block detection with something like:

```diff
diff --git a/src/lib.rs b/src/lib.rs
index d3379a45..4117590f 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -387,6 +387,16 @@ fn enclose_in_main_block(s: &str, config: &Config) -> String {
     let block_len = formatted
         .snippet
         .rfind('}')
+        .map(
+            // trim and trailing whitespace we added, e.g. if we format to
+            // fn main { some_func(); }
+            |i| {
+                i - formatted.snippet[..i]
+                    .chars()
+                    .take_while(|c| *c == ' ')
+                    .count()
+            },
+        )
         .unwrap_or_else(|| formatted.snippet.len());
 
     // It's possible that `block_len < FN_MAIN_PREFIX.len()`. This can happen if the code block was
```

File: tests/target/issue-6832.rs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// rustfmt-format_code_in_doc_comments: true
+// rustfmt-fn_single_line: true
+// rustfmt-error_on_unformatted: true
```
Comment (matthewhughes934) [inline tests/target/issue-6832.rs]: `error_on_unformatted` isn't necessary here, the test will fail if the code needs reformatting

```suggestion
```

Comment (caelyreth) [reply tests/target/issue-6832.rs]: Yes, after my local testing, this one is unnecessary. I'll delete this line laterly.

File: src/lib.rs (hunk 1 above)
Comment (caelyreth) [reply src/lib.rs]: I agree with your idea that we may consider unifying things in one place.

But, the patches you proposed do not fix the problem. It seems you forgot to call `.rev()` so that the iterator can see from the end (trailing).

Also, I think the `take_while()` here has less code readability since it is not straightforward to see what it is. I may prefer the following for better semantics.

```rust
.map(|i| {
    // A single-line wrapper becomes `fn main() { stmt }`. Trim the
    // wrapper's space before its closing brace from the extracted body.
    let body = &formatted.snippet[..i];
    if body.contains('\n') {
        i
    } else {
        body.strip_suffix(' ').map_or(i, str::len)
    }
})
```

or, even simpler, if we broaden our logic to all cases instead of only the collapsed one-line:

```rust
.map(|i| formatted.snippet[..i].strip_suffix(' ').map_or(i, str::len))
```

Both passed all test cases. What do you think?

Comment (matthewhughes934) [reply src/lib.rs]: > But, the patches you proposed do not fix the problem. It seems you forgot to call .rev() so that the iterator can see from the end (trailing).

lol, not sure how I failed the copy-pasta so well, I had `.rev` locally with tests passing :sweat_smile: 

> What do you think?

The one-line approach seems reasonable to me

File: src/lib.rs [hunk 3]
```diff
@@ -387,6 +387,7 @@ fn format_code_block(
     let block_len = formatted
         .snippet
         .rfind('}')
+        .map(|i| formatted.snippet[..i].strip_suffix(' ').map_or(i, str::len))
```
Comment (matthewhughes934) [inline src/lib.rs:391]: *[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/6851/1faf40273eb16f42da0f5e76f635f5348feca914..8097bc550ecd566a06bafb563e78ee75d588a827)*

I'm wondering if it's clear why we do this, or if it's still worth an extra comment 🤔 

```suggestion
        // trim and trailing whitespace we added, e.g. if we format to fn main { some_func(); }
        .map(|i| formatted.snippet[..i].strip_suffix(' ').map_or(i, str::len))
```

**Review by matthewhughes934 — APPROVED** [review]

**Comment by afurm** [comment]
The fix handles the single-line wrapped case narrowly. Is there a scenario where `format_code_block` is called with a multi-line function inside a doc comment that also has `fn_single_line = true`? Would the same wrapper-space leak affect that case?

**Comment by utakotoba** [comment]
> The fix handles the single-line wrapped case narrowly. Is there a scenario where `format_code_block` is called with a multi-line function inside a doc comment that also has `fn_single_line = true`? Would the same wrapper-space leak affect that case?

Yes. `format_code_block()` is used for any Rust fenced code block in a **doc comment**, not just already-single-line snippets.

The leak depends on the final wrapped formatted shape, **not** the original input shape. So if a multi-line function in a doc comment is reformatted such that the dummy wrapper collapses to `fn main() { ... }`, the same wrapper-space leak can occur.

The current fix is applied at extraction time, so it also covers that case. If the wrapped output stays multi-line, `strip_suffix(' ')` is a no-op.

File: src/lib.rs (hunk 3 above)
Comment (utakotoba) [reply src/lib.rs:391]: Yes, I think it is good to note this.

File: src/lib.rs [hunk 4]
```diff
@@ -387,6 +387,8 @@ fn format_code_block(
     let block_len = formatted
         .snippet
         .rfind('}')
+        // trim and trailing whitespace we added, e.g. if we format to fn main { some_func(); }
```
Comment (ytmimi) [inline src/lib.rs]: Could you mention that `fn main { some_func(); }` could only happen when `fn_single_line=true` is used. It took me a second to figure out what the code was doing so maybe a diagram would help?


```suggestion
        // trim whitespace when using `fn_single_line=true`
        // ```
        //                        | `block_len` starts at the index of `}`
        // fn main { some_func(); }
        //                      | remove whitespace to move it to the index of `;`
        // ```
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/6851/1faf40273eb16f42da0f5e76f635f5348feca914..0bab52c1722e01f2b2db94507f2c9188fbd053e9)*

File: src/lib.rs [hunk 5]
```diff
@@ -387,6 +387,8 @@ fn format_code_block(
...
+        .map(|i| formatted.snippet[..i].strip_suffix(' ').map_or(i, str::len))
```
Comment (ytmimi) [inline src/lib.rs]: Might be a little more flexible to use `char::is_whitespace`
```suggestion
        .map(|i| formatted.snippet[..i].strip_suffix(char::is_whitespace).map_or(i, str::len))
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/6851/1faf40273eb16f42da0f5e76f635f5348feca914..0bab52c1722e01f2b2db94507f2c9188fbd053e9)*

File: tests/target/issue-6832.rs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+// rustfmt-format_code_in_doc_comments: true
+// rustfmt-fn_single_line: true
+
+/// ```rust
+/// let font = String::from("hello");
+/// ```
+fn main() {}
```
Comment (ytmimi) [inline tests/target/issue-6832.rs]: Can we add another test case that uses multi-byte unicode characters.

````rust
/// ```rust
/// let font = String::from("hello 👋😁");
/// ```
fn unicode_chars() {}
````

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/6851/1faf40273eb16f42da0f5e76f635f5348feca914..0bab52c1722e01f2b2db94507f2c9188fbd053e9)*

File: src/lib.rs (hunk 4 above)
Comment (caelyreth) [reply src/lib.rs]: Yes, I think this is reasonable. Thx.

File: src/lib.rs (hunk 5 above)
Comment (caelyreth) [reply src/lib.rs]: Yes, that's great. I didn't even realize before that we could use `char::is_whitespace` as a pattern here.

**Comment by caelyreth** [comment]
@ytmimi Thanks for your insightful review!

**Comment by Ciznia** [comment]
Any news on when it will be merged ?
